File: MasteringSwift/MasteringSwift/chapters/chapter19/basic/BasicPerformance.swift
//
//  BasicPerformance.swift
//  MasteringSwift
//
//  Capítulo 19 — Rendimiento y optimización.
//  Nivel BÁSICO: Eficiente vs ineficiente — bucle vs built-in, valores cacheados.
//

import Foundation

func runBasicPerformanceExamples() {
    let numbers = [1, 2, 3, 4, 5]

    var sumLoop = 0
    for n in numbers { sumLoop += n }
    print("Sum (loop): \(sumLoop)")

    let sumBuiltIn = numbers.reduce(0, +)
    print("Sum (built-in reduce): \(sumBuiltIn)")

    let list = (0..<100).map { $0 }
    let factor = 2
    var sumRepeated = 0
    for n in list { sumRepeated += n * factor }
    let doubled = list.map { $0 * factor }
    var sumCached = 0
    for n in doubled { sumCached += n }
    print("Sum repeated calc: \(sumRepeated), cached: \(sumCached)")
}

File: MasteringDart/lib/chapters/chapter19/basic/basic_performance.dart
// ignore_for_file: avoid_print
//
//  basic_performance.dart
//  MasteringDart
//
//  Capítulo 19 — Rendimiento y optimización.
//  Nivel BÁSICO: Eficiente vs ineficiente — bucle vs built-in, valores en caché.
//
//  Ejecutar con: dart run lib/chapters/chapter19/basic/basic_performance.dart

void runBasicPerformanceExamples() {
  const numbers = [1, 2, 3, 4, 5];

  var sumLoop = 0;
  for (final n in numbers) sumLoop += n;
  print('Sum (loop): $sumLoop');

  final sumBuiltIn = numbers.reduce((a, b) => a + b);
  print('Sum (built-in reduce): $sumBuiltIn');

  final list = List.generate(100, (i) => i);
  const factor = 2;
  var sumRepeated = 0;
  for (final n in list) sumRepeated += n * factor;
  final doubled = list.map((n) => n * factor).toList();
  var sumCached = 0;
  for (final n in doubled) sumCached += n;
  print('Sum repeated calc: $sumRepeated, cached: $sumCached');
}

void main() {
  runBasicPerformanceExamples();
}
